### PR TITLE
Use config-reload container as an init container

### LINKED ISSUE #868: On start, CasC configuration is reloaded way too often
### Is your feature request related to a problem? Please describe

In our setup I experience the following issue: Upon restart of Jenkins, the config-reload sidecar sends up to 28 configuration-reload requests to the Jenkins process. We have 12 casc-configmaps.
So, for _each_ ConfigMap, the config file is being created, a reload-request sent to Jenkins, which reloads all configuration files.

So the first config file will be reloaded 12 times (In our case probably even more because of some other issues).

The main issue that I have with this, that it takes time and is unnecessary work.

### Describe the solution you'd like

I propose to run kiwigrid additionally as an initContainer, so that all casc configuration files are in place before Jenkins even starts.
Just run the kiwigrid with `METHOD="LIST"` once and that's it. That makes restarting a lot easier and faster, I hope.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dominykas**
### What does this PR do?

- Fixes #868 

This should make the startup routine of Jenkins more stable and have fewer race conditions. When the existing implementation of the sidecar detects configmap/secret changes, it calls an endpoint on Jenkins to reload CasC, however since the sidecar and the Jenkins begin booting up at the same time, this introduces several potential race conditions:

- the endpoint to reload CasC may be called before Jenkins has started up and result in various errors (e.g. CSRF problems)
- the endpoint to reload CasC may be called while Jenkins is initializing original CasC (unverified, but in theory possible - there is some evidence that this might have happened, although not concrete proof)
- the config-reload sidecar might be rewriting files while Jenkins is reading them, resulting in partial, inconsistent configuration which in turn results in a crash (we have definitely experienced something that can be explained by this)

Introducing an init container with METHOD=LIST copies the necessary files into place before Jenkins has started booting up, preventing the above race conditions.

I verified this approach in practice, and whereby Jenkins used to crash 3-4 times (at least) before finally starting up, it does now start up consistently without crashes.

There still is scope for certain race conditions due to the fact that Helm applies changes all at once, however they should be unlikely, as various delays in StatefulSet restarting the Pod and the cluster taking time to start up the Pod.

If you modified files in the `./charts/jenkins/` directory, please also include the following:

```[tasklist]
### Submitter checklist
- [x] I bumped the "version" key in `./charts/jenkins/Chart.yml`.
- [x] I added a new changelog entry to `./charts/jenkins/CHANGELOG.md`.
- [x] I followed the [technical requirements](https://github.com/jenkinsci/helm-charts/blob/main/CONTRIBUTING.md#technical-requirements).
```

### Special notes for your reviewer

- It might be easier to review this commit-by-commit.
- Should I hide the init container behind a flag and disable it by default? I feel like that should not be necessary, but at the same time - I don't know how large the user base is for this and what kinds of configurations they might have - some might consider this a breaking change?

### CODE REVIEW

File: charts/jenkins/templates/_helpers.tpl [hunk 1]
```diff
@@ -480,3 +480,60 @@ Create the HTTP port for interacting with the controller
...
+      value: "{{ template "jenkins.fullname" $root }}-jenkins-config"
+    - name: FOLDER
+      value: "{{ $root.Values.controller.sidecars.configAutoReload.folder }}"
+    - name: NAMESPACE
+      value: '{{ $root.Values.controller.sidecars.configAutoReload.searchNamespace | default (include "jenkins.namespace" $root) }}'
+{{- if $method }}
+    - name: METHOD
+      value: "{{ $method }}"
+{{- end }}
+{{- if $root.Values.controller.sidecars.configAutoReload.sleepTime }}
```
Comment (torstenwalter) [inline charts/jenkins/templates/_helpers.tpl:513]: Should we use an else if here to ensure that we can never end up with two `METHOD` values?

**Review by torstenwalter — COMMENTED** [review]
This sounds like a good approach to fix the race condition. It's just important to keep the reload container also as a sidecar to allow live reloads, while Jenkins is running. That's the case here.

@dominykas appreciate the commit descriptions. That makes it easy to review.

File: charts/jenkins/templates/_helpers.tpl (hunk 1 above)
Comment (dominykas) [reply charts/jenkins/templates/_helpers.tpl]: Does https://github.com/jenkinsci/helm-charts/pull/926/commits/9d3a61e45d29fe5e7aa876e2cf8e086f392837f1 work?

**Review by torstenwalter — APPROVED** [review]

**Comment by dominykas** [comment]
Thanks!

**Comment by timja** [comment]
> Thanks!

unfortunately we need to revert this: https://github.com/jenkinsci/helm-charts/pull/930

see https://matrix.to/#/!JLUOInpEYmxJIYXlzs:matrix.org/$6_PcNdByVjmF8pqChuA0E3E6lGyhNjFYiHBYzc_Mnxc?via=libera.chat&via=gitter.im&via=matrix.org

**Comment by dominykas** [comment]
Yes, I saw that, I think it does not crash, though? It only logs the warnings?

A fix should be easy - the URL should simply not be set in the env.

I'll PR that later today.

**Comment by dominykas** [comment]
Ah, I see there's an issue with people who were using `METHOD` as well (they could now set `sleepTime` instead). This could have been considered a breaking change, yeah... but I'll see if I can address that too in a non-breaking fashion.

**Comment by timja** [comment]
> Yes, I saw that, I think it does not crash, though? It only logs the warnings?
> 
> A fix should be easy - the URL should simply not be set in the env.
> 
> I'll PR that later today.

It would not start: https://github.com/jenkinsci/helm-charts/pull/930#issuecomment-1757832910

**Pull request merged**